qiskit/qasm2/parse.py

```diff
@@ -0,0 +1,405 @@
+    Measure,
+    QuantumCircuit,
+    QuantumRegister,
+    Qubit,
+    Reset,
+    library as lib,
+)
+
+# This is the same C-extension problems as described in the `__init__.py` disable near the
+# `_qasm2` import.
+from qiskit._qasm2 import (  # pylint: disable=no-name-in-module
+    OpCode,
+    UnaryOpCode,
+    BinaryOpCode,
+    CustomClassical,
+    ExprConstant,
+    ExprArgument,
+    ExprUnary,
+    ExprBinary,
+    ExprCustom,
+)
+from .exceptions import QASM2ParseError
+
+# Constructors of the form `*params -> Gate` for the special 'qelib1.inc' include.  This is
+# essentially a pre-parsed state for the file as given in the arXiv paper defining OQ2.
+QELIB1 = (
+    lib.U3Gate,
+    lib.U2Gate,
+    lib.U1Gate,
+    lib.CXGate,
+    # IGate in Terra < 0.24 is defined as a single-cycle delay, so is not strictly an identity.
+    # Instead we use a true no-op stand-in.
+    lambda: lib.UGate(0, 0, 0),
+    lib.XGate,
+    lib.YGate,
+    lib.ZGate,
+    lib.HGate,
+    lib.SGate,
+    lib.SdgGate,
+    lib.TGate,
+    lib.TdgGate,
+    lib.RXGate,
+    lib.RYGate,
+    lib.RZGate,
+    lib.CZGate,
+    lib.CYGate,
+    lib.CHGate,
+    lib.CCXGate,
+    lib.CRZGate,
+    lib.CU1Gate,
+    lib.CU3Gate,
+)
+
+
+@dataclasses.dataclass(frozen=True)
+class CustomInstruction:
+    """Information about a custom instruction that should be defined during the parse.
+
+    The ``name``, ``num_params`` and ``num_qubits`` fields are self-explanatory.  The
+    ``constructor`` field should be a callable object with signature ``*args -> Instruction``, where
+    each of the ``num_params`` ``args`` is a floating-point value.  Most of the built-in Qiskit gate
+    classes have this form.
+
+    There is a final ``builtin`` field.  This is optional, and if set true will cause the
+    instruction to be defined and available within the parsing, even if there is no definition in
+    any included OpenQASM 2 file.
+    """
+
+    name: str
+    num_params: int
+    num_qubits: int
+    # This should be `(float*) -> Instruction`, but the older version of Sphinx we're constrained to
+    # use in the Python 3.9 docs build chokes on it, so relax the hint.
+    constructor: Callable[..., Instruction]
+    builtin: bool = False
+
+
+def _generate_delay(time: float):
+    # This wrapper is just to ensure that the correct type of exception gets emitted; it would be
+    # unnecessarily spaghetti-ish to check every emitted instruction in Rust for integer
+    # compatibility, but only if its name is `delay` _and_ its constructor wraps Qiskit's `Delay`.
+    if int(time) != time:
+        raise QASM2ParseError("the custom 'delay' instruction can only accept an integer parameter")
+    return Delay(int(time), unit="dt")
+
+
+class _U0Gate(Gate):
+    def __init__(self, count):
+        if int(count) != count:
+            raise QASM2ParseError("the number of single-qubit delay lengths must be an integer")
+        super().__init__("u0", 1, [count])
```

Comment:
the parser gives a float in `operands`, that triggers `TypeError: can't multiply sequence by non-int of type 'float'` at decomposition time.
```suggestion
        super().__init__("u0", 1, [int(count)])
```

It might be good to test something like `qiskit.qasm2.loads("qreg q[1]; u0(1) q[0];", custom_instructions=qiskit.qasm2.LEGACY_CUSTOM_INSTRUCTIONS).decompose()` to catch this.

Reply:
Thanks - done in fdc6a9dbb.